src/sentry/migrations/0165_create_notificationsetting_table.py

```diff
@@ -0,0 +1,176 @@
+                                    0
+                                ),
+                                "default",
+                            ),
+                            (
+                                sentry.models.notificationsetting.NotificationSettingOptionValues(
+                                    10
+                                ),
+                                "off",
+                            ),
+                            (
+                                sentry.models.notificationsetting.NotificationSettingOptionValues(
+                                    20
+                                ),
+                                "on",
+                            ),
+                            (
+                                sentry.models.notificationsetting.NotificationSettingOptionValues(
+                                    30
+                                ),
+                                "subscribe_only",
+                            ),
+                            (
+                                sentry.models.notificationsetting.NotificationSettingOptionValues(
+                                    40
+                                ),
+                                "committed_only",
+                            ),
+                        ]
+                    ),
+                ),
+                (
+                    "organization",
+                    sentry.db.models.fields.foreignkey.FlexibleForeignKey(
+                        null=True,
+                        on_delete=django.db.models.deletion.CASCADE,
+                        to="sentry.Organization",
+                    ),
+                ),
+                (
+                    "project",
+                    sentry.db.models.fields.foreignkey.FlexibleForeignKey(
+                        null=True, on_delete=django.db.models.deletion.CASCADE, to="sentry.Project"
+                    ),
+                ),
+                (
+                    "team",
+                    sentry.db.models.fields.foreignkey.FlexibleForeignKey(
+                        null=True, on_delete=django.db.models.deletion.CASCADE, to="sentry.Team"
+                    ),
+                ),
+                (
+                    "user",
+                    sentry.db.models.fields.foreignkey.FlexibleForeignKey(
+                        null=True,
+                        on_delete=django.db.models.deletion.CASCADE,
+                        to=settings.AUTH_USER_MODEL,
+                    ),
+                ),
+            ],
+            options={
+                "db_table": "sentry_notificationsetting",
+            },
+        ),
+        migrations.AlterUniqueTogether(
+            name="notificationsetting",
+            unique_together={
+                ("organization", "project", "team", "type", "provider"),
+                ("organization", "project", "user", "type", "provider"),
```

Comment:
Just a note about unique constraints - if any of these values are null, then we will allow duplicates:

```
class TestUnique(Model):
    __core__ = False

    field_1 = models.IntegerField(null=True)
    field_2 = models.IntegerField(null=True)
    field_3 = models.IntegerField(null=True)
    field_4 = models.IntegerField(null=True)

    class Meta:
        app_label = "sentry"
        db_table = "sentry_testunique"
        unique_together = (("field_1", "field_2", "field_3", "field_4"),)
```
Then creating rows works as expected with all cols:
```
TestUnique.objects.create(field_1=1, field_2=2, field_3=3, field_4=4)
<TestUnique at 0x116b199b0: id=3>

TestUnique.objects.create(field_1=1, field_2=2, field_3=3, field_4=4)
IntegrityError: UniqueViolation('duplicate key value violates unique constraint "sentry_testunique_field_1_field_2_field_3_field_4_0d9050a1_uniq"\nDETAIL:  Key (field_1, field_2, field_3, field_4)=(1, 2, 3, 4) already exists.\n',)
SQL: INSERT INTO "sentry_testunique" ("field_1", "field_2", "field_3", "field_4") VALUES (%s, %s, %s, %s) RETURNING "sentry_testunique"."id"
```
But with partial cols
```
TestUnique.objects.create(field_1=1, field_2=None, field_3=3, field_4=4)
<TestUnique at 0x1168829b0: id=5>
TestUnique.objects.create(field_1=1, field_2=None, field_3=3, field_4=4)
<TestUnique at 0x116aad2e8: id=6>
TestUnique.objects.create(field_1=1, field_2=None, field_3=3, field_4=4)
<TestUnique at 0x115ddbdd8: id=7>
```

Feels like this can happen based on the constraints you have?

https://stackoverflow.com/questions/8289100/create-unique-constraint-with-null-columns might be helpful

Reply:
Here's what we're thinking as a replacement for `unique_together`: 
```
migrations.RunSQL(
    """
    CREATE UNIQUE INDEX sentry_notificationsetting_org_user_uniq 
    ON sentry_notificationsetting (organization_id, user_id, provider, type) 
    WHERE project_id IN NULL AND team_id IS NULL AND user_id IS NOT NULL;
    """
),
migrations.RunSQL(
    """
    CREATE UNIQUE INDEX sentry_notificationsetting_org_team_uniq 
    ON sentry_notificationsetting (organization_id, team_id, provider, type) 
    WHERE project_id IN NULL AND user_id IS NULL AND team_id IS NOT NULL;
    """
),
migrations.RunSQL(
    """
    CREATE UNIQUE INDEX sentry_notificationsetting_org_proj_user_uniq 
    ON sentry_notificationsetting (organization_id, project_id, user_id, provider, type) 
    WHERE team_id IS NULL AND user_id IS NOT NULL;
    """
),
migrations.RunSQL(
    """
    CREATE UNIQUE INDEX sentry_notificationsetting_org_proj_team_uniq 
    ON sentry_notificationsetting (organization_id, project_id, team_id, provider, type) 
    WHERE user_id IS NULL AND team_id IS NOT NULL;
    """
),
migrations.RunSQL(
    """
	CREATE UNIQUE INDEX sentry_notificationsetting_user_uniq 
	ON sentry_notificationsetting (user_id, provider, type) 
	WHERE organization_id IS NULL AND project_id IS NULL AND user_id IS NOT NULL;
    """
),	
```
we'd create unique constraints with partial indexes for every valid combination. Is this the right approach?